test/e2e/tests/vault-corruption/vault-corruption.spec.ts

```diff
@@ -0,0 +1,340 @@
+   * background page has reloaded and the UI is available again.
+   *
+   * @param driver - The WebDriver instance.
+   * @param script - The script to break the DB that will be executed in the
+   * background page for MV2 or offscreen page for MV3.
+   * @returns The initial first account's address.
+   */
+  async function onboardThenCorruptVault(driver: Driver, script: string) {
+    const initialWindow = await driver.driver.getWindowHandle();
+
+    // open a spare tab so the browser doesn't exit once we `reload()` the
+    // extension process, as doing so will close all UI tabs we have open when
+    // we do -- and that will close the whole browser 😱
+    await driver.openNewPage('about:blank');
+
+    await onboard(driver);
+
+    const homePage = new HomePage(driver);
+    homePage.check_pageIsLoaded();
+
+    const headerNavbar = new HeaderNavbar(driver);
+    const firstAddress = await getFirstAddress(driver, headerNavbar);
+    await headerNavbar.lockMetaMask();
+
+    // use the home page to destroy the vault
+    await driver.executeAsyncScript(script);
+
+    // the previous tab we were using is now closed, so we need to tell Selenium
+    // to switch back to the other page (required for Chrome)
+    await driver.switchToWindow(initialWindow);
+
+    // get a new tab ready to use (required for Firefox)
+    await driver.openNewPage('about:blank');
+
+    // wait for the background page to reload
+    await waitForVaultRestorePage(driver);
+
+    return firstAddress;
+  }
+
+  /**
+   * Returns the vault recovery button.
+   *
+   * @param driver - The WebDriver instance.
+   */
+  async function getVaultRecoveryButton(driver: Driver) {
+    return await driver.waitForSelector('#critical-error-button', {
+      // The button is disabled for a timeout period as a precaution, so we
+      // need to wait for it to be enabled before continuing.
+      state: 'enabled',
+    });
+  }
+
+  /**
+   * Click the recovery/reset button then confirm or dismiss the action.
+   *
+   * @param driver - The WebDriver instance.
+   * @param confirm - Whether to confirm the action or not.
+   */
+  async function clickRecover(driver: Driver, confirm: boolean) {
+    // click the Recovery/Reset button
+    const recoveryButton = await getVaultRecoveryButton(driver);
+    await recoveryButton.click();
+
+    // Confirm we want to recover/reset.
+    const prompt = await driver.driver.switchTo().alert();
+    if (confirm) {
+      await prompt.accept();
+    } else {
+      await prompt.dismiss();
+    }
+
+    // the button should be disabled while the recovery process is in progress,
+    // and enabled if the user dismissed the prompt
+    const isEnabled = await recoveryButton.isEnabled();
+    assert.equal(
+      isEnabled,
+      !confirm,
+      confirm
+        ? 'Recovery button should be disabled when prompt is accepted'
+        : 'Recovery button should be enabled when prompt is dismissed',
+    );
```

Comment:
mm I think this assertion could lead to race conditions, where the element is briefly in an undesired state while the page is rendering (enabled/disabled) - ie the element is already in the DOM with its default state and then changes to whichever state needs to be.

To prevent this kind of flakiness maybe we could do this:

`driver.wait(until.elementIsEnabled(recoveryButton), timeout)`

or this:

`driver.wait(until.elementIsDisabled(recoveryButton), timeout)`

depending on the `confirm` condition. It's not exactly the same, but with this, we are accounting for brief intermediate states that can happen during the e2e fast actions.

What do you think?

Reply:
I see, then I guess we can leave it like this, as this is a critical flow and if the button is an undesired state it flags a bug :pray: I think it's fine like this then, and if we see in the future that this adds flakiness, we can either see if that can be fixed on the app side, or tweak it here.

Reply:
@Gudahtt the operations (`accept`/`dismiss` and set the button `disabled` state), are _always synchronous_ on the browser side. Selenium can't check the state in between the two actions.

Reply:
Ah sorry, I see, I was confused. I was thinking we were in React and this button was starting out as disabled. But it's starting as enabled (of course, it was just clicked), and we're not using React so it is synchronous as you said. You're right, this is fine.